### DISCUSSION TITLE
firmware debugging procedure

### DISCUSSION THREAD

**Discussion by gusido** [pine64/OpenPineBuds#66] [Q&A] [answered]
Hi, 

First, thanks for an awesome project!

Second - is there any way to debug the firmware? trace? printf? anything alone these lines to help figure out what's going on on the device in real time?  

The only reference I could find of "debug" is in the readme concerning debugging baud rate, but i have not idea what to make of it.

Thanks!

**Comment by shymega** [comment]
There's not, really. @Ralim reverse engineered the buds for flashing, but I think debugging is a long way off. The typical way you'd do it is via UART, which only works whilst the buds are in the case. 

However, it could be worth exposing a BLE service - once licensing is done, and CMake is merged, plus refactoring - to print log messages. We'd need to abstract a logger over that.

**Comment by Ralim** [comment, marked as answer]
Hello 👋🏼 

So the buds spew debugging info out the uart all the time. However if you connect them to the charger they go into sleep mode.
So what I do is  put a small piece of tape over the 5V pogo pin so they dont see the charger, then I can use the charger as a uart connection to debug output.

This way printf can be used in debugging.

There is no out-of-box debug at the moment.

When I'm adjusting audio related stuff, i put one in and then have the other on the uart to get logs (since they stay sync'ed)

**Reply by gusido** [reply]
Thanks!! This is very helpful, excited to try it out.

**Reply by nicka101** [reply]
@Ralim Can you clarify which pin is the 5v, and whether any changes are needed (to [config/open_source/target.mk#L20](https://github.com/pine64/OpenPineBuds/blob/main/config/open_source/target.mk#L20) for example) to get this working?

I've tried covering each of the 4 pins individually on both earphones, and no ttyUSB device ever appears

**Reply by Ralim** [reply]
The case will show up as two USB<->UART adaptors even without buds installed. Covering 5V is only required to stop them going to sleep (thinking they are in the case). (Not sure on 5V pin off the top of my head, will need to check).

No changes to makefile were required, though I did sometimes change log levels to tradeoff information.

**Reply by nicka101** [reply]
For me the case identifies as "QinHeng Electronics USB Dual_Serial"  in lsusb and presents two ACM devices for flashing the buds, but I've never seen any ttyUSB devices such as referenced in ./uart_log.sh.

Do I need a specific kernel module loaded? Seems like my system has picked cdc_acm

**Reply by shymega** [reply]
No kernel module is needed, unless you've done some major tweaks. Check for `/dev/ttyACM*` devices.

**Reply by nicka101** [reply]
@shymega yeah, I see the ttyACM devices, but no ttyUSB

Here's my dmesg output (2 ttyACM devices, as expected)
![image](https://github.com/pine64/OpenPineBuds/assets/1499871/ea8f12f6-4c81-459e-bbba-c61b1b7b2e5e)

But no ttyUSB devices (and attempting to use the ttyACM ones as uart doesn't produce anything usable)
![image](https://github.com/pine64/OpenPineBuds/assets/1499871/0c4081c6-f2c6-4e1f-a904-f5572464b59c)

In case its relevant, I've tried both Fedora (with kernel 6.8.7) and Manjaro (with kernel 6.6.25) as the host system, and both load cdc_acm and detect the ttyACM devices, but never ttyUSB

**Reply by shymega** [reply]
@nicka101 It's practically the same thing. I see `ttyACM` too on NixOS. As it's recognised as a dual serial USB device, `ACM0`/`ACM`1 correspond to each bud individually in the case.

I'm currently rewriting `bestool`, but eventually you'll be able to use `bestool` to read the serial logs from the buds. It won't be automatic at this time. However, for now, I hope I've given enough context to understand that `ttyACM0/1` should work for you. 

The `uart_log.sh` script is inconsistent with the other scripts. Some use `ttyUSB` and some use `ttyACM`.

**Reply by nicka101** [reply]
Ah thanks, I've got it figured out now, as it turns out the default configuration for minicom isn't compatible with the buds, and the invocation in the `uart_log.sh` script doesn't actually change the tty device or baud rate.

Perhaps we could add this `minirc.dfl` to the repo (it uses the left bud):
```
pu port             /dev/ttyACM1
pu baudrate         2000000
pu bits             8
pu parity           N
pu stopbits         1
```
or change `uart_log.sh` to something like:
```
#!/usr/bin/env sh

num=$(find /dev -name 'ttyACM*' | rev | cut -c 1)
echo "$num"
sudo minicom -D /dev/ttyACM$num -b 2000000

```
to be more inline with the other scripts and consistent across systems

**Reply by nicka101** [reply]
@shymega That said, I've opened #99 to address it, and it should also improve compatibility with systems with other serial devices, as well as give you the choice of which bud to debug

**Reply by shymega** [reply]
The problem is, that we don't want to assume it's `ttyACM*` on every system. Some systems, like macOS, use different devnodes. I would rather avoid setting it as a default (a la #99) and instead rely on `bestool` when I finish refactoring.

Thank you for the PR, though. But I think I'd rather just add a change for the baud rate, and leave it at that...

**Reply by nicka101** [reply]
You already assume that its ttyACM in a bunch of places and ttyUSB in others, surely its beneficial to have a functional script for the majority of environments (#99 should work anywhere with a linux kernel above 2.5, though not sure about Mac or other BSD derivatives), rather than a non-functional script  until an indeterminate future point when bestool is refactored to provide the same?

By all means change it to improve Mac/BSD compatibility, but it seems strange to keep the script present and also broken, and potentially puts off new contributors

**Reply by shymega** [reply]
I suppose we could merge it. I'm just erring on the side of caution. I'm also keeping in mind the rewrite, which will handle these things.

**Reply by nicka101** [reply]
Fair enough, do whatever you think is best. I'm just not sure that there's any real downside to (at least mostly) fixing the script now, even if it is going to be replaced with something better/more reliable later

**Reply by shymega** [reply]
It looks like @Ralim has merged it. I'm mostly cautious with merges, and this is one of my weaknesses. Thank you for taking the initiative to make a PR though - action is by doing. Very much appreciated. And now your work is in the repo!
